### PR TITLE
[TCFMM-14] Add `x_to_y` and `y_to_x` tests

### PR DESCRIPTION
**Pull request by dcastro**
[//]: # (This is a template of a pull request template.)
[//]: # (You should modify it considering specifics of a particular repository and put it there.)
[//]: # (Comments like this are meta-comments, they shouldn't be present in the final template.)
[//]: # (Comments starting with '

[//]: # (Here you can add a link to the corresponding issue tracker, e. g. https://issues.serokell.io/issue/AD-)
[//]: # (For GitHub/GitLab issues it is better to use just hash symbol or exclamation mark as it is more resistant to repo movements)
[//]: # (In this case please also prefix the link with "Resolves" keyword)
[//]: # (See https://docs.gitlab.com/ee/user/project/issues/managing_issues.html#closing-issues-automatically)
[//]: # (See https://help.github.com/en/github/managing-your-work-on-github/closing-issues-using-keywords)
## Related issue(s)

Resolves part of TCFMM-14

#### Related changes (conditional)

- Tests
  - [x] If I added new functionality, I added tests covering it.
  - [x] If I fixed a bug, I added a regression test to prevent the bug from
        silently reappearing again.

[//]: # (Add more docs here if you have them in the repository)
- Documentation
  - [x] I checked whether I should update the docs and did so if necessary:
    - [Specification](../tree/master/docs/specification.md)
    - [README](../tree/master/README.md)
    - Haddock

#### Stylistic guide (mandatory)

[//]: # (Update link to style guide if necesary or remove if it's not present)

- [x] My commits comply with [the policy used in Serokell](https://www.notion.so/serokell/Where-and-how-to-commit-your-work-58f8973a4b3142c8abbd2e6fd5b3a08e).
- [x] My code complies with the [style guide](../tree/master/docs/code-style.md).

### CODE REVIEW

File: haskell/test/Test/Util.hs [hunk 1]
```diff
@@ -86,7 +86,10 @@ originateSegCFMM
   => TokenType -> TokenType -> Storage
   -> m (ContractHandler Parameter Storage)
 originateSegCFMM xTokenType yTokenType storage = do
-  originateSimple "Segmented CFMM" storage $ segCFMMContract xTokenType yTokenType
+  cfmm <- originateSimple "Segmented CFMM" storage $ segCFMMContract xTokenType yTokenType
+  -- Add some slots to the buffers to make the tests more meaningful.
+  call cfmm (Call @"Increase_observation_count") 10
```
Comment (pasqu4le) [inline haskell/test/Test/Util.hs:93]: :thinking: couldn't this be done by modifying the storage instead?

File: ligo/errors.mligo [hunk 2]
```diff
@@ -149,6 +149,9 @@
 (* Some problem in cumulative seconds per liquidity computation. *)
 [@inline] let internal_spl_cumulative_err = 326n
 
+(* Flip of `seconds_per_liquidity_outside` failed. (This is an invariant of the contract). *)
+[@inline] let flip_seconds_per_liquidity_outside_err = 327n
```
Comment (pasqu4le) [inline ligo/errors.mligo:153]: shouldn't this be an internal error, aka only happen due to a bug in the contract?

If so, I guess we should keep the convention to include `internal` in its name, I think

**Review by pasqu4le — COMMENTED** [review]
Really well done, I didn't expect the required changes to be so extensive :clap: 

Will review it in more detail, but at a first look all makes sense to me :ok_hand:

File: ligo/errors.mligo (hunk 2 above)
Comment (dcastro) [reply ligo/errors.mligo:153]: Aye good point! I just copied `flip_fee_growth_outside_err` and did not notice that, I'll rename both errors then.

Comment (dcastro) [reply ligo/errors.mligo]: Done

File: haskell/test/Test/Util.hs (hunk 1 above)
Comment (Martoon-00) [reply haskell/test/Test/Util.hs:93]: Maybe better not perform this in the basic origination function? 
1. This hinders writing tests on observation logic itself (I will have to account for this `+10` even if I wanted to test another scenario)
2. Now mostly all tests will fail if something goes wrong in `Increase_observation_count` logic, probably we'd rather make only few tests be affected by that.

Comment (dcastro) [reply haskell/test/Test/Util.hs:93]: Good point, I'll move this away from `originateSegCFMM` then

File: haskell/test/Test/Invariants.hs [hunk 3]
```diff
@@ -8,22 +8,90 @@ module Test.Invariants
...
+      Accumulators
+        { aSeconds = aSeconds tc1 + aSeconds tc2
+        , aTickCumulative = aTickCumulative tc1 + aTickCumulative tc2
+        , aFeeGrowth = aFeeGrowth tc1 + aFeeGrowth tc2
+        , aSecondsPerLiquidity = aSecondsPerLiquidity tc1 + aSecondsPerLiquidity tc2
+        }
+
+-- | Invariants:
+-- 1. The contract always has enough balance to liquidite all positions (and pay any fees due).
+checkBalanceInvariants :: (HasCallStack, MonadEmulated caps base m) => ContractHandler Parameter Storage -> Storage -> m ()
```
Comment (pasqu4le) [inline haskell/test/Test/Invariants.hs:72]: Minor nitpick, but I suppose that this and the one in `checkAccumulatorsInvariants` don't need to be "listed".

Unless you want to count them all in their order in `checkAllInvariants`, but in that case the numbers should be changed.

**Review by pasqu4le — APPROVED** [review]
All checks out to me :+1:

File: haskell/test/Test/Invariants.hs (hunk 3 above)
Comment (dcastro) [reply haskell/test/Test/Invariants.hs:72]: Done

**Pull request merged**